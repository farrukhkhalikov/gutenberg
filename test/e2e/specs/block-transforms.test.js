/**
 * External dependencies
 */
import path from 'path';
import { some } from 'lodash';

/**
 * Internal dependencies
 */
import {
	newPost,
	getEditedPostContent,
	setPostContent,
	getAllBlocks,
	selectBlockByClientId,
} from '../support/utils';
import {
	getFileBaseNames,
	readFixtureFile,
} from '../../support/utils';

const BLOCK_SWITCHER_SELECTOR = '.editor-block-toolbar .editor-block-switcher';

const getAvailableBlockTransforms = async () => {
	return await page.evaluate( () => {
		return Array.from(
			document.querySelectorAll(
				'.editor-block-types-list .editor-block-types-list__list-item button'
			)
		).map(
			( button ) => {
				return button.getAttribute( 'aria-label' );
			}
		);
	} );
};

const hasBlockSwitcher = async () => {
	return await page.evaluate( ( blockSwitcherSelector ) => {
		return !! document.querySelector( blockSwitcherSelector );
	}, BLOCK_SWITCHER_SELECTOR );
};

const getProcessedFileBase = async ( fixturesDir, fileBase ) => {
	if ( some(
		[ 'deprecated', 'subhead', 'text-columns' ],
		( partialFileName ) => fileBase.includes( partialFileName )
	) ) {
		return {
			expectedDeprecatedBlock: true,
		};
	}
	const parsedBlockObject = JSON.parse(
		readFixtureFile( fixturesDir, fileBase + '.parsed.json' )
	);
	if ( parsedBlockObject.blockName === null ) {
		return {
			expectedInvalidBlock: true,
		};
	}
	const content = readFixtureFile( fixturesDir, fileBase + '.html' );
	await setPostContent( content );
	const blocks = await getAllBlocks();
	const normalizedMarkup = await getEditedPostContent();
	await selectBlockByClientId( blocks[ 0 ].clientId );
	let availableTransforms;
	if ( await hasBlockSwitcher() ) {
		await page.click( BLOCK_SWITCHER_SELECTOR );
		availableTransforms = await getAvailableBlockTransforms(
			blocks[ 0 ].clientId
		);
	}

	return {
		normalizedMarkup,
		availableTransforms,
	};
};

describe( 'test transforms', () => {
	const fixturesDir = path.join(
		__dirname, '..', '..', 'integration', 'full-content', 'fixtures'
	);

	const fileBasenames = getFileBaseNames( fixturesDir );

	const blocksParsed = {};
	beforeAll( async () => {
		await newPost();
		for ( const fileBase of fileBasenames ) {
			blocksParsed[ fileBase ] = await getProcessedFileBase( fixturesDir, fileBase );
		}
		expect( console ).toHaveErroredWith(
			'Failed to load resource: the server responded with a status of 404 (Not Found)'
		);
	} );

	it( 'all transforms work', async () => {
		expect( JSON.stringify( blocksParsed ) ).toMatchSnapshot();
	} );
} );
