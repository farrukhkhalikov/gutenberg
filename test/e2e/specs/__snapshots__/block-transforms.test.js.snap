// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test transforms all transforms work 1`] = `"{\\"core__4-invalid-starting-letter\\":{},\\"core__archives\\":{},\\"core__archives__showPostCounts\\":{},\\"core__audio\\":{\\"availableTransforms\\":[\\"File\\"]},\\"core__block\\":{},\\"core__button__center\\":{\\"availableTransforms\\":[]},\\"core__categories\\":{},\\"core__code\\":{\\"availableTransforms\\":[\\"Preformatted\\"]},\\"core__column\\":{\\"availableTransforms\\":[\\"Heading\\",\\"List\\",\\"Quote\\",\\"Preformatted\\",\\"Verse\\"]},\\"core__columns\\":{\\"availableTransforms\\":[\\"Heading\\",\\"List\\",\\"Quote\\",\\"Preformatted\\",\\"Verse\\"]},\\"core__cover-image\\":{\\"availableTransforms\\":[\\"Heading\\",\\"Image\\"]},\\"core__embed\\":{},\\"core__file__new-window\\":{},\\"core__file__no-download-button\\":{},\\"core__file__no-text-link\\":{},\\"core__freeform\\":{},\\"core__freeform__undelimited\\":{},\\"core__gallery\\":{\\"availableTransforms\\":[\\"Image\\"]},\\"core__gallery__columns\\":{\\"availableTransforms\\":[\\"Image\\"]},\\"core__heading__h2-em\\":{\\"availableTransforms\\":[\\"Quote\\",\\"Cover Image\\",\\"Paragraph\\"]},\\"core__heading__h2\\":{\\"availableTransforms\\":[\\"Quote\\",\\"Cover Image\\",\\"Paragraph\\"]},\\"core__html\\":{},\\"core__image\\":{\\"availableTransforms\\":[\\"Gallery\\",\\"Cover Image\\",\\"File\\"]},\\"core__image__attachment-link\\":{\\"availableTransforms\\":[\\"Gallery\\",\\"Cover Image\\",\\"File\\"]},\\"core__image__center-caption\\":{\\"availableTransforms\\":[\\"Gallery\\",\\"Cover Image\\",\\"File\\"]},\\"core__image__custom-link\\":{\\"availableTransforms\\":[\\"Gallery\\",\\"Cover Image\\",\\"File\\"]},\\"core__image__media-link\\":{\\"availableTransforms\\":[\\"Gallery\\",\\"Cover Image\\",\\"File\\"]},\\"core__invalid-Capitals\\":{},\\"core__invalid-special\\":{},\\"core__latest-comments\\":{},\\"core__latest-posts\\":{},\\"core__latest-posts__displayPostDate\\":{},\\"core__list__ul\\":{\\"availableTransforms\\":[\\"Paragraph\\",\\"Quote\\"]},\\"core__more\\":{},\\"core__more__custom-text-teaser\\":{},\\"core__nextpage\\":{},\\"core__paragraph__align-right\\":{\\"availableTransforms\\":[\\"Heading\\",\\"List\\",\\"Quote\\",\\"Preformatted\\",\\"Verse\\"]},\\"core__paragraph__deprecated\\":{\\"expectedDeprecatedBlock\\":true},\\"core__preformatted\\":{\\"availableTransforms\\":[\\"Paragraph\\"]},\\"core__pullquote\\":{\\"availableTransforms\\":[]},\\"core__pullquote__multi-paragraph\\":{\\"availableTransforms\\":[]},\\"core__quote__style-1\\":{\\"availableTransforms\\":[\\"List\\",\\"Paragraph\\",\\"Heading\\"]},\\"core__quote__style-2\\":{\\"availableTransforms\\":[\\"List\\",\\"Paragraph\\",\\"Heading\\"]},\\"core__separator\\":{\\"availableTransforms\\":[]},\\"core__shortcode\\":{},\\"core__spacer\\":{},\\"core__subhead\\":{\\"expectedDeprecatedBlock\\":true},\\"core__table\\":{},\\"core__text-columns\\":{\\"expectedDeprecatedBlock\\":true},\\"core__text__converts-to-paragraph\\":{\\"availableTransforms\\":[\\"Heading\\",\\"List\\",\\"Quote\\",\\"Preformatted\\",\\"Verse\\"]},\\"core__verse\\":{\\"availableTransforms\\":[\\"Paragraph\\"]},\\"core__video\\":{\\"availableTransforms\\":[\\"File\\"]}}"`;
